library/alloc/src/vec/mod.rs

```diff
@@ -2058,6 +2058,42 @@ impl<T, A: Allocator> Vec<T, A> {
         }
     }
 
+    /// Removes and returns the last element in a vector if the predicate
+    /// returns `true`, or [`None`] if the predicate returns false or the vector
+    /// is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(vec_pop_if)]
+    ///
+    /// let mut vec = vec![1, 2, 3, 4];
+    /// let pred = |x: &mut i32| *x % 2 == 0;
+    ///
+    /// assert_eq!(vec.pop_if(pred), Some(4));
+    /// assert_eq!(vec, [1, 2, 3]);
+    /// assert_eq!(vec.pop_if(pred), None);
+    /// ```
+    #[unstable(feature = "vec_pop_if", issue = "122741")]
+    pub fn pop_if<F>(&mut self, f: F) -> Option<T>
+    where
+        F: FnOnce(&mut T) -> bool,
+    {
+        if self.len == 0 {
+            return None;
+        }
+
+        // SAFETY: the vector is not empty
+        let mut last = unsafe { ptr::read(self.as_ptr().add(self.len() - 1)) };
+
+        if f(&mut last) {
```

Comment:
This needs some type of drop guard because `f` might panic and that would've run `T`'s `drop` without adjusting the length of the vec.

Reply:
Makes sense, is there an example of that pattern I could reference?

Reply:
You can probably just wrap the last value in a `ManuallyDrop`, that should do it too.